File: shifai-android/app/src/main/java/com/shifai/infrastructure/PerformanceMonitor.kt
package com.shifai.infrastructure

import android.os.SystemClock
import android.util.Log
import java.util.concurrent.ConcurrentHashMap

/**
 * Performance Monitoring — mirrors iOS PerformanceMonitor.swift
 * Uses os.log equivalent (android.util.Log) with NFR threshold checks.
 */
object PerformanceMonitor {

    private const val TAG = "ShifAI.Perf"

    // NFR thresholds (milliseconds)
    object Thresholds {
        const val COLD_START_MS = 4000L      // < 4s
        const val WARM_START_MS = 2000L      // < 2s
        const val TRANSITION_MS = 300L       // < 300ms
        const val ML_INFERENCE_MS = 500L     // < 500ms
        const val SYNC_PUSH_MS = 5000L       // < 5s
        const val PDF_GENERATION_MS = 3000L  // < 3s
    }

    private val markers = ConcurrentHashMap<String, Long>()

    /**
     * Start a timing measurement.
     */
    fun startMeasure(label: String) {
        markers[label] = SystemClock.elapsedRealtime()
        Log.d(TAG, "⏱ START: $label")
    }

    /**
     * End a timing measurement. Returns elapsed ms.
     * Logs warnings if NFR thresholds are exceeded.
     */
    fun endMeasure(label: String): Long? {
        val start = markers.remove(label) ?: return null
        val elapsed = SystemClock.elapsedRealtime() - start

        // Check against thresholds
        when {
            label.contains("cold_start") && elapsed > Thresholds.COLD_START_MS ->
                Log.w(TAG, "⚠️ Cold start ${elapsed}ms > ${Thresholds.COLD_START_MS}ms target")
            label.contains("warm_start") && elapsed > Thresholds.WARM_START_MS ->
                Log.w(TAG, "⚠️ Warm start ${elapsed}ms > ${Thresholds.WARM_START_MS}ms target")
            label.contains("transition") && elapsed > Thresholds.TRANSITION_MS ->
                Log.w(TAG, "⚠️ Transition ${elapsed}ms > ${Thresholds.TRANSITION_MS}ms target")
            label.contains("ml_inference") && elapsed > Thresholds.ML_INFERENCE_MS ->
                Log.w(TAG, "⚠️ ML inference ${elapsed}ms > ${Thresholds.ML_INFERENCE_MS}ms target")
            label.contains("sync") && elapsed > Thresholds.SYNC_PUSH_MS ->
                Log.w(TAG, "⚠️ Sync ${elapsed}ms > ${Thresholds.SYNC_PUSH_MS}ms target")
            label.contains("pdf") && elapsed > Thresholds.PDF_GENERATION_MS ->
                Log.w(TAG, "⚠️ PDF gen ${elapsed}ms > ${Thresholds.PDF_GENERATION_MS}ms target")
        }

        Log.d(TAG, "⏱ END: $label = ${elapsed}ms")
        return elapsed
    }

    /**
     * Log memory usage.
     */
    fun logMemory() {
        val runtime = Runtime.getRuntime()
        val used = (runtime.totalMemory() - runtime.freeMemory()) / (1024 * 1024)
        val max = runtime.maxMemory() / (1024 * 1024)
        Log.d(TAG, "📊 Memory: ${used}MB / ${max}MB")
    }

    /**
     * Convenience: measure a block of code.
     */
    inline fun <T> measure(label: String, block: () -> T): T {
        startMeasure(label)
        val result = block()
        endMeasure(label)
        return result
    }
}

File: shifai-ios/ShifAI/Infrastructure/PerformanceMonitor.swift
import Foundation
import os.log

// MARK: - Performance Monitor (S10-1)
// NFR-P1→P7: Cold start <4s, warm <1s, transitions <300ms, ML <150ms, battery <5%

final class PerformanceMonitor {

    static let shared = PerformanceMonitor()
    private let logger = Logger(subsystem: "com.shifai", category: "Performance")

    // MARK: - Timing

    private var markers: [String: CFAbsoluteTime] = [:]

    func startMeasure(_ label: String) {
        markers[label] = CFAbsoluteTimeGetCurrent()
    }

    func endMeasure(_ label: String) -> TimeInterval? {
        guard let start = markers.removeValue(forKey: label) else { return nil }
        let elapsed = CFAbsoluteTimeGetCurrent() - start
        let ms = elapsed * 1000

        // Log with threshold warnings
        switch label {
        case let l where l.contains("cold_start"):
            if ms > 4000 { logger.warning("⚠️ Cold start \(ms, privacy: .public)ms > 4000ms target") }
            else { logger.info("✅ Cold start \(ms, privacy: .public)ms") }
        case let l where l.contains("warm_start"):
            if ms > 1000 { logger.warning("⚠️ Warm start \(ms, privacy: .public)ms > 1000ms target") }
        case let l where l.contains("transition"):
            if ms > 300 { logger.warning("⚠️ Transition \(ms, privacy: .public)ms > 300ms target") }
        case let l where l.contains("ml_inference"):
            if ms > 150 { logger.warning("⚠️ ML inference \(ms, privacy: .public)ms > 150ms target") }
        case let l where l.contains("sync"):
            if ms > 2000 { logger.warning("⚠️ Sync \(ms, privacy: .public)ms > 2000ms target") }
        default:
            logger.info("📊 \(label, privacy: .public): \(ms, privacy: .public)ms")
        }

        return elapsed
    }

    // MARK: - Memory

    func logMemoryUsage() {
        var info = mach_task_basic_info()
        var count = mach_msg_type_number_t(MemoryLayout<mach_task_basic_info>.size) / 4
        let result = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                task_info(mach_task_self_, task_flavor_t(MACH_TASK_BASIC_INFO), $0, &count)
            }
        }
        if result == KERN_SUCCESS {
            let mb = Double(info.resident_size) / 1024 / 1024
            logger.info("💾 Memory: \(mb, privacy: .public)MB")
        }
    }

    // MARK: - NFR Thresholds

    struct Thresholds {
        static let coldStartMs: Double = 4000
        static let warmStartMs: Double = 1000
        static let transitionMs: Double = 300
        static let mlInferenceMs: Double = 150
        static let syncMs: Double = 2000
        static let batteryPercentDay: Double = 5.0
    }
}
